### ISSUE TITLE
More Legend Options

### ISSUE THREAD

**Issue by levimann24** [ScottPlot/ScottPlot#1805]
## Feature Suggestion

**Feature description:** 
I have a few ideas for legend features, there is a chance I have missed a couple of these in the code, so just let me know if that's the case.
1. Ability to make the legend horizontal, and appear outside of the plot.

      Here is an example from Matplotlib showing the legend both outside and horizontal.
      ![image](https://user-images.githubusercontent.com/65826749/165540239-0f323a56-70dd-417f-9b06-a2bdcde3521c.png)

2. Ability for horizontal legends to automatically update layout when there are too many series to fit in a legend that has the same width as the plot. 

     Here is an example from OxyPlot showing this. In this case soon as the series count gets to four, the next series in the legend shows up in a second row. 
      ![image](https://user-images.githubusercontent.com/65826749/165541014-d8d3ddbb-e897-45bc-92c7-a4f791a45e2f.png)

3. Ability to change the color of the legend to match the current plot style.

Let me know what y'all think!

**Comment by BambOoxX** [comment]
Hi. I think this has been brought up in some previous issues/discussion, and that would be grezt, but at the moment AFAIK it is not possible to draw anything outside of the plotting area since everything is rendered there. So basically you would be taking space for plottables

**Comment by levimann24** [comment]
He @BambOoxX, thanks for the quick reply. Is there a way to shrink the area for plottables and add the legend there? Regardless of being able to do that now, it would be nice to have a horizonal option. Once you have several series on the plot, the legend tends to be in the way.

**Comment by BambOoxX** [comment]
@levimann24 Yes, when you have a lot of legend entries, the legend starts to take a lot of space.

Depending on how you actually build the plot can really change the behavior I believe.
For instance if you use it as a standalone plot you could resize the image at will and compensate for the shrinking du to the legend.
However if you use it in a gui control, the overall size is fixed and leaving more space for the legend will only reduce the plotting area.
Basically, there is no good solution to this problem.

My way of handling this was to create the "detach legend" capability in winforms so that one can display the legend in a separate window from the plots. It can be improved, but that's another solution.

Regarding the horizontal legend, it should be feasible, @swharden can telle his mind on this, and I can have a look at how it could be implemented.

**Comment by levimann24** [comment]
@BambOoxX Thanks for all of the info. 

I am working with Winforms and I have used the "detach legend," great work on that by the way, it works great when I have a single plot, but the application I am working on is using multiple plots in the same window, and it can get difficult to keep track of each of the legends, still doable. 

I am thinking about a temporary solution, on my end, to hide/show the legend depending on the mouse position compared to legend position.  

I have also never contributed to a public project, so I am also willing to jump in and help out!

**Comment by BambOoxX** [comment]
> I am working with Winforms and I have used the "detach legend," great work on that by the way,

@levimann24 Thanks for the kind words. 

> it works great when I have a single plot, but the application I am working on is using multiple plots in the same window, and it can get difficult to keep track of each of the legends, still doable. 

Technically, I think it could be feasible to dock the detached legend into some fixed placeholder so that it appears next to a plot. I've never thought about this as it still may take some space in the overall window, but that could fit your demand. 

> I am thinking about a temporary solution, on my end, to hide/show the legend depending on the mouse position compared to legend position.  

In fact, you could capture the `MouseEnter` event of the `FormsPlot` control so that you can perform this dynamic hide /show of the legends. 
Depending on what you want to achieve exactly, you could also capture the legend that is under the mouse and highlight the frame of the associated `FormsPlot` or else. 

> I have also never contributed to a public project, so I am also willing to jump in and help out! 

This project is actually the biggest I've contributed to myself. @swharden is very welcoming, so I'm sure you'll be able to help easily.

**Comment by swharden** [comment]
Hi @levimann24 and @BambOoxX, lots of good discussion here!

It would be possible to redesign the legend system to allow it to nestle on one side of the plot, but to be honest that would be a lot of complex work (the layout system is pretty advanced), and a lot of extra code I wouldn't be very excited to maintain. Part of the complexity is that determining how large of an area you need depends on measuring strings, and that can be different on different operating systems and vary depending on available system fonts. This may get better in a future version of ScottPlot, but for now this is what makes this task more difficult than one would initially expect.

It may not be obvious how simple the existing legend is: It's just a _picture_ that gets placed on top of the plot at the end after everything else is rendered. 

If you consider that the legend is just a picture, it shouldn't be too intimidating to create a static method that accepts an array of items in the legend and returns the legend as a picture. Different methods (or arguments) could customize orientation, rows/columns, etc.

```cs
// how to get legend items from a plot
LegendItem[] items = plt.GetPlottables()
                        .Where(x => x.GetLegendItems() != null)
                        .SelectMany(x => x.GetLegendItems())
                        .Where(x => !string.IsNullOrEmpty(x.label))
                        .ToArray();
```

If you create static methods which are useful for generating legend images, feel free to share them here and we can place them somewhere in ScottPlot so others can benefit from those legend image generators too!

As a hack if you want to try to draw the legend image in a custom location on the image you can make some extra space on one of the edges by calling `plt.Layout()` and using the custom arguments.

**Comment by levimann24** [comment]
@swharden Thanks for the info. I will play around with is and see if I can come up with a temporary solution!
